Intermitent NameNotFoundException for java:global JNDI name during application start

**Describe the bug**  
A NameNotFoundException may occur intermittently during application start if the application initialization has multiple threads that may concurrently perform a JNDI lookup of something in the java:global and/or java:app name contexts. The failure appears as follows:

Caused by: javax.naming.NameNotFoundException: javax.naming.NameNotFoundException: java:global/jdbc/DataSourceDef
    at com.ibm.ws.jndi.url.contexts.javacolon.internal.JavaURLContext.lookup(JavaURLContext.java:357)
    at com.ibm.ws.jndi.url.contexts.javacolon.internal.JavaURLContext.lookup(JavaURLContext.java:372)
    at org.apache.aries.jndi.DelegateContext.lookup(DelegateContext.java:149)
    at java.naming/javax.naming.InitialContext.lookup(InitialContext.java:409)
    at java.naming/javax.naming.InitialContext.doLookup(InitialContext.java:282)
    at io.openliberty.data.internal.persistence.service.DelegatingResourceFactory.createResource(DelegatingResourceFactory.java:59)
    at com.ibm.wsspi.persistence.internal.DatabaseStoreImpl.createPersistenceServiceUnit(DatabaseStoreImpl.java:286)


**Steps to Reproduce**  
Not easily reproducible. An application must have initialization that runs concurrent threads that perform JNDI lookups in java:global and/or java:app.  

**Expected behavior**  
JNDI lookups of java:global and java:app should work consistently every time an application starts.

**Diagnostic information:**  
 - OpenLiberty Version: [24.0.03]
 - Affected feature(s) [jndi-1.0]
 - Java Version: [all]

**Additional context**  
None

--- a/dev/com.ibm.ws.injection/src/com/ibm/ws/injectionengine/osgi/internal/OSGiInjectionScopeData.java
+++ b/dev/com.ibm.ws.injection/src/com/ibm/ws/injectionengine/osgi/internal/OSGiInjectionScopeData.java
@@ -1,10 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2019 IBM Corporation and others.
+ * Copyright (c) 2011, 2024 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -23,6 +23,8 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
@@ -129,6 +131,15 @@ public class OSGiInjectionScopeData extends InjectionScopeData implements Deferr
      * @see #deferredReferenceDataEnabled
      */
     private Map<DeferredReferenceData, Boolean> deferredReferenceDatas;
+
+    /**
+     * The result of {@link #processDeferredReferenceData} for the current list of reference
+     * contexts that are registered for deferred processing if a non-java:comp request is made.
+     * This field is initialized lazily and cleared when no longer needed.
+     *
+     * @see #deferredReferenceDatas
+     */
+    private CompletableFuture<Boolean> deferredReferenceDatasFuture;
 
     public OSGiInjectionScopeData(J2EEName j2eeName, NamingConstants.JavaColonNamespace namespace, OSGiInjectionScopeData parent, ReentrantReadWriteLock nonCompEnvLock) {
         super(j2eeName);
@@ -515,6 +526,8 @@ public class OSGiInjectionScopeData extends InjectionScopeData implements Deferr
         if (parent != null && deferredReferenceDatas != null) {
             parent.removeDeferredReferenceData(this);
             deferredReferenceDatas = null;
+            deferredReferenceDatasFuture.complete(false);
+            deferredReferenceDatasFuture = null;
         }
     }
 
@@ -528,6 +541,7 @@ public class OSGiInjectionScopeData extends InjectionScopeData implements Deferr
 
         if (deferredReferenceDatas == null) {
             deferredReferenceDatas = new LinkedHashMap<DeferredReferenceData, Boolean>();
+            deferredReferenceDatasFuture = new CompletableFuture<Boolean>();
             if (parent != null && deferredReferenceDataEnabled) {
                 parent.addDeferredReferenceData(this);
             }
@@ -563,18 +577,17 @@ public class OSGiInjectionScopeData extends InjectionScopeData implements Deferr
      */
     @Override
     public boolean processDeferredReferenceData() {
-        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
-            Tr.entry(tc, "processDeferredReferenceData", "this=" + this);
+        final boolean isTraceOn = TraceComponent.isAnyTracingEnabled();
+        if (isTraceOn && tc.isEntryEnabled()) {
+            Tr.entry(tc, "processDeferredReferenceData", "this=" + this + " future=" + deferredReferenceDatasFuture);
         }
 
         Map<DeferredReferenceData, Boolean> deferredReferenceDatas;
+        CompletableFuture<Boolean> currentFuture;
         synchronized (this) {
             deferredReferenceDatas = this.deferredReferenceDatas;
             this.deferredReferenceDatas = null;
-
-            if (parent != null) {
-                parent.removeDeferredReferenceData(this);
-            }
+            currentFuture = deferredReferenceDatasFuture;
         }
 
         boolean any = false;
@@ -589,12 +602,38 @@ public class OSGiInjectionScopeData extends InjectionScopeData implements Deferr
                     // (erroneous or conflicting metadata).  Any exception that
                     // is thrown will be rethrown by ReferenceContext.process
                     // when the component is actually used.
-                    ex.getClass(); // findbugs
+                    if (isTraceOn && tc.isDebugEnabled())
+                        Tr.debug(tc, "ignoring : " + ex);
                 }
+            }
+
+            synchronized (this) {
+                // Removed from parent after processing completes so concurrent access of parent
+                // will block on the parent future until all child processing completes.
+                if (parent != null) {
+                    parent.removeDeferredReferenceData(this);
+                }
+
+                // Unblock any concurrent access and clear the future as is no longer required.
+                if (deferredReferenceDatasFuture != null) {
+                    deferredReferenceDatasFuture.complete(any);
+                    deferredReferenceDatasFuture = null;
+                }
+            }
+        } else if (currentFuture != null) {
+            try {
+                // Wait a reasonable amount of time for deferred processing to complete
+                if (isTraceOn && tc.isDebugEnabled())
+                    Tr.debug(tc, "waiting up to 60 seconds for completion of " + currentFuture);
+                any = currentFuture.get(60, TimeUnit.SECONDS);
+            } catch (Exception e) {
+                if (isTraceOn && tc.isDebugEnabled())
+                    Tr.debug(tc, "processDeferredReferenceData failed to complete; assuming something was processed");
+                any = true;
             }
         }
 
-        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+        if (isTraceOn && tc.isEntryEnabled()) {
             Tr.exit(tc, "processDeferredReferenceData", any);
         }
         return any;

--- a/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java
+++ b/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java
@@ -175,7 +175,7 @@ public class FutureEMBuilder extends CompletableFuture<EntityManagerBuilder> {
                     }
                 } catch (NamingException x) {
                     if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-                        Tr.debug(this, tc, "Ignoring lookup failure for " + javaCompName + " : " + x);
+                        Tr.debug(this, tc, javaCompName + " is not available in JNDI");
                 }
             }
         } finally {

--- a/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java
+++ b/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java
@@ -16,6 +16,7 @@ import java.util.HashSet;
 import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 
 import javax.naming.InitialContext;
 import javax.naming.NamingException;
@@ -25,6 +26,7 @@ import com.ibm.websphere.csi.J2EEName;
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
 import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.ws.ffdc.FFDCFilter;
 import com.ibm.ws.ffdc.annotation.FFDCIgnore;
 import com.ibm.ws.runtime.metadata.ComponentMetaData;
 import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
@@ -150,6 +152,9 @@ public class FutureEMBuilder extends CompletableFuture<EntityManagerBuilder> {
                                         resourceName, metadataIdentifier, application, module, component, entityTypes);
 
                 } catch (NamingException x) {
+                    FFDCFilter.processException(x, this.getClass().getName() + ".createEMBuilder", "155", this, new Object[] { (switchMetadata ? repoMetadata : extMetadata) });
+                    throw new CompletionException("Unable to find " + dataStore + " from " +
+                                                  (switchMetadata ? repoMetadata : extMetadata).getJ2EEName(), x);
                 }
             } else if (!DataExtension.DEFAULT_DATA_STORE.equals(resourceName)) {
                 // Check for resource references and persistence unit references where java:comp/env/ is omitted:
@@ -169,6 +174,8 @@ public class FutureEMBuilder extends CompletableFuture<EntityManagerBuilder> {
                         resourceName = javaCompName;
                     }
                 } catch (NamingException x) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                        Tr.debug(this, tc, "Ignoring lookup failure for " + javaCompName + " : " + x);
                 }
             }
         } finally {
@@ -214,12 +221,12 @@ public class FutureEMBuilder extends CompletableFuture<EntityManagerBuilder> {
                                             (application == null ? 4 : application.length()) +
                                             (module == null ? 4 : module.length()) +
                                             (component == null ? 4 : component.length())) //
-                                                            .append("FutureEMBuilder@") //
-                                                            .append(Integer.toHexString(hashCode())) //
-                                                            .append(":").append(dataStore) //
-                                                            .append(' ').append(application) //
-                                                            .append('#').append(module) //
-                                                            .append('#').append(component);
+                        .append("FutureEMBuilder@") //
+                        .append(Integer.toHexString(hashCode())) //
+                        .append(":").append(dataStore) //
+                        .append(' ').append(application) //
+                        .append('#').append(module) //
+                        .append('#').append(component);
         return b.toString();
     }
 }

